File: tests/python/llamaindex/test_prototype.py
"""Conformance test: prototype instrumentation for LlamaIndex."""

import json
import os

from opentelemetry import trace
from opentelemetry._logs import get_logger_provider

from otel_setup import flush_and_shutdown, setup_otel

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"] + "/v1"

_prototype_tracer = trace.get_tracer("gen_ai.prototype")


def run_chat_prototype(llm, request_model, request_temperature):
    """Scenario: basic chat completion with prototype instrumentation."""
    from llama_index.core.llms import ChatMessage, MessageRole

    print("  [chat] basic chat completion (prototype)")
    with _prototype_tracer.start_as_current_span("chat gpt-4o-mini") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        span.set_attribute("gen_ai.request.temperature", request_temperature)
        user_content = "Say hello."
        resp = llm.chat([ChatMessage(role=MessageRole.USER, content=user_content)])
        raw = getattr(resp, "raw", None)
        if raw:
            if getattr(raw, "model", None):
                span.set_attribute("gen_ai.response.model", raw.model)
            if getattr(raw, "id", None):
                span.set_attribute("gen_ai.response.id", raw.id)
            if getattr(raw, "choices", None):
                span.set_attribute("gen_ai.response.finish_reasons", [c.finish_reason for c in raw.choices])
            if getattr(raw, "usage", None) and raw.usage:
                span.set_attribute("gen_ai.usage.input_tokens", raw.usage.prompt_tokens)
                span.set_attribute("gen_ai.usage.output_tokens", raw.usage.completion_tokens)

        # Emit inference operation details event
        event_attrs = {
            "gen_ai.operation.name": "chat",
            "gen_ai.request.model": request_model,
            "gen_ai.input.messages": json.dumps([
                {"role": "user", "parts": [{"type": "text", "content": user_content}]}
            ]),
            "gen_ai.output.messages": json.dumps([
                {
                    "role": "assistant",
                    "parts": [{"type": "text", "content": str(resp)}],
                    "finish_reason": raw.choices[0].finish_reason if raw and getattr(raw, "choices", None) else None,
                }
            ]),
        }
        if raw:
            if getattr(raw, "model", None):
                event_attrs["gen_ai.response.model"] = raw.model
            if getattr(raw, "id", None):
                event_attrs["gen_ai.response.id"] = raw.id
            if getattr(raw, "choices", None):
                event_attrs["gen_ai.response.finish_reasons"] = [c.finish_reason for c in raw.choices]
            if getattr(raw, "usage", None) and raw.usage:
                event_attrs["gen_ai.usage.input_tokens"] = raw.usage.prompt_tokens
                event_attrs["gen_ai.usage.output_tokens"] = raw.usage.completion_tokens
        get_logger_provider().get_logger("gen_ai.prototype").emit(
            event_name="gen_ai.client.inference.operation.details",
            body="Inference operation details",
            attributes=event_attrs,
        )

        print(f"    -> {str(resp)[:60]}")


def run_chat_streaming_prototype(llm, request_model, request_temperature):
    """Scenario: streaming chat completion with prototype instrumentation."""
    from llama_index.core.llms import ChatMessage, MessageRole

    print("  [chat_streaming] streaming chat completion (prototype)")
    with _prototype_tracer.start_as_current_span("chat gpt-4o-mini") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        span.set_attribute("gen_ai.request.temperature", request_temperature)
        text = ""
        stream_resp = llm.stream_chat(
            [ChatMessage(role=MessageRole.USER, content="Tell me a joke.")]
        )
        for token in stream_resp:
            text += token.delta
        raw = getattr(stream_resp, "raw", None)
        if raw:
            if getattr(raw, "model", None):
                span.set_attribute("gen_ai.response.model", raw.model)
            if getattr(raw, "id", None):
                span.set_attribute("gen_ai.response.id", raw.id)
        print(f"    -> {text[:60]}")


def run_agent_prototype(llm, request_model, request_temperature):
    """Scenario: agent with tool calling and prototype instrumentation."""
    print("  [chat_tool_call] agent with tool calling (prototype)")
    from llama_index.core.tools import FunctionTool

    def get_weather(location: str) -> str:
        """Get the current weather for a location."""
        return "Sunny, 72°F"

    weather_tool = FunctionTool.from_defaults(fn=get_weather)
    tool_definition = {
        "name": weather_tool.metadata.name,
        "description": weather_tool.metadata.description,
        "fn_schema": weather_tool.metadata.fn_schema.model_json_schema(),
    }
    with _prototype_tracer.start_as_current_span("chat gpt-4o-mini") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        span.set_attribute("gen_ai.request.temperature", request_temperature)
        span.set_attribute("gen_ai.tool.definitions", json.dumps([tool_definition]))

        response = llm.predict_and_call(
            tools=[weather_tool],
            user_msg="What's the weather in Seattle?",
            verbose=False,
        )
        print(f"    -> {str(response)[:60]}")


def run_embeddings_prototype():
    """Scenario: embedding generation with prototype instrumentation."""
    print("  [embeddings] embedding generation (prototype)")
    from llama_index.embeddings.openai import OpenAIEmbedding

    request_model = "text-embedding-3-small"
    embed_model = OpenAIEmbedding(
        model_name=request_model,
        api_base=MOCK_BASE_URL,
        api_key="mock-key",
    )
    with _prototype_tracer.start_as_current_span("embeddings text-embedding-3-small") as span:
        span.set_attribute("gen_ai.operation.name", "embeddings")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        result = embed_model.get_text_embedding("Hello, world!")
        print(f"    -> embedding dim: {len(result)}")


def main():
    print("=== Prototype: LlamaIndex Conformance Test ===")

    tp, lp, mp = setup_otel()

    from llama_index.llms.openai import OpenAI as LlamaOpenAI

    request_model = "gpt-4o-mini"
    request_temperature = 0.1
    llm = LlamaOpenAI(
        model=request_model,
        temperature=request_temperature,
        api_base=MOCK_BASE_URL,
        api_key="mock-key",
    )

    run_chat_prototype(llm, request_model, request_temperature)
    run_chat_streaming_prototype(llm, request_model, request_temperature)
    run_agent_prototype(llm, request_model, request_temperature)
    run_embeddings_prototype()

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()

File: tests/js/llamaindex/test_prototype.ts
/**
 * Conformance test: Prototype instrumentation for LlamaIndex.
 */
import { trace } from "@opentelemetry/api";
import { logs, SeverityNumber } from "@opentelemetry/api-logs";
import { flushAndShutdownOtel, setupOtel } from "../otel";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_BASE_URL = nodeProcess.env.MOCK_LLM_URL! + "/v1";
const tracer = trace.getTracer("gen_ai.prototype");

async function main() {
  console.log("=== Prototype: LlamaIndex JS Conformance Test ===");
  const otel = setupOtel();

  const { OpenAI, OpenAIEmbedding } = await import("@llamaindex/openai");

  const chatModel = "gpt-4o-mini";
  const requestTopP = 1.0;
  const llm = new OpenAI({
    model: chatModel,
    topP: requestTopP,
    additionalSessionOptions: { baseURL: MOCK_BASE_URL },
    apiKey: "mock-key",
  });

  // Scenario: chat
  console.log("  [chat] basic chat completion (prototype)");
  await tracer.startActiveSpan("chat gpt-4o-mini", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", chatModel);
    span.setAttribute("gen_ai.request.top_p", requestTopP);
    const userMessage = "Say hello.";
    const resp = await llm.chat({
      messages: [{ role: "user", content: userMessage }],
    });
    const raw = resp.raw as any;
    if (raw?.model) span.setAttribute("gen_ai.response.model", raw.model);
    if (raw?.id) span.setAttribute("gen_ai.response.id", raw.id);
    if (raw?.choices?.[0]?.finish_reason) {
      span.setAttribute("gen_ai.response.finish_reasons", [raw.choices[0].finish_reason]);
    }
    if (raw?.usage) {
      span.setAttribute("gen_ai.usage.input_tokens", raw.usage.prompt_tokens);
      span.setAttribute("gen_ai.usage.output_tokens", raw.usage.completion_tokens);
    }

    // Emit inference operation details event
    logs.getLogger("gen_ai.prototype").emit({
      severityNumber: SeverityNumber.INFO,
      eventName: "gen_ai.client.inference.operation.details",
      body: "Inference operation details",
      attributes: {
        "gen_ai.operation.name": "chat",
        "gen_ai.request.model": chatModel,
        "gen_ai.response.id": raw?.id,
        "gen_ai.response.model": raw?.model,
        "gen_ai.response.finish_reasons": raw?.choices?.[0]?.finish_reason ? [raw.choices[0].finish_reason] : undefined,
        "gen_ai.usage.input_tokens": raw?.usage?.prompt_tokens,
        "gen_ai.usage.output_tokens": raw?.usage?.completion_tokens,
        "gen_ai.input.messages": JSON.stringify([
          { role: "user", parts: [{ type: "text", content: userMessage }] }
        ]),
        "gen_ai.output.messages": JSON.stringify([{
          role: "assistant",
          parts: [{ type: "text", content: resp.message.content.toString() }],
          finish_reason: raw?.choices?.[0]?.finish_reason,
        }]),
      },
    });

    console.log(`    -> ${resp.message.content.toString().slice(0, 60)}`);
    span.end();
  });

  // Scenario: chat with tool calling
  console.log("  [chat_tool_call] chat with tool calling (prototype)");
  await tracer.startActiveSpan("chat gpt-4o-mini", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", chatModel);
    const requestTool = {
      type: "function" as const,
      function: {
        name: "get_weather",
        description: "Get the current weather",
        parameters: {
          type: "object",
          properties: {
            location: { type: "string", description: "City name" },
          },
          required: ["location"],
        },
      },
    };
    span.setAttribute("gen_ai.tool.definitions", JSON.stringify([requestTool]));
    const resp = await llm.chat({
      messages: [{ role: "user", content: "What's the weather in Seattle?" }],
      additionalChatOptions: { tools: [requestTool] },
    });
    const raw = resp.raw as any;
    if (raw?.model) span.setAttribute("gen_ai.response.model", raw.model);
    if (raw?.id) span.setAttribute("gen_ai.response.id", raw.id);
    if (raw?.choices?.[0]?.finish_reason) {
      span.setAttribute("gen_ai.response.finish_reasons", [raw.choices[0].finish_reason]);
    }
    if (raw?.usage) {
      span.setAttribute("gen_ai.usage.input_tokens", raw.usage.prompt_tokens);
      span.setAttribute("gen_ai.usage.output_tokens", raw.usage.completion_tokens);
    }
    if (raw?.choices?.[0]?.message?.tool_calls?.length) {
      console.log(`    -> tool_call: ${raw.choices[0].message.tool_calls[0].function.name}`);
    } else {
      console.log(`    -> ${resp.message.content.toString().slice(0, 60)}`);
    }
    span.end();
  });

  // Scenario: embeddings
  console.log("  [embeddings] embedding generation (prototype)");
  await tracer.startActiveSpan("embeddings text-embedding-3-small", async (span) => {
    const requestModel = "text-embedding-3-small";
    span.setAttribute("gen_ai.operation.name", "embeddings");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const embed = new OpenAIEmbedding({
      model: requestModel,
      apiKey: "mock-key",
      additionalSessionOptions: { baseURL: MOCK_BASE_URL },
    });
    const result = await embed.getTextEmbedding("Hello, world!");
    console.log(`    -> embedding dim: ${result.length}`);
    span.end();
  });

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });
